---
title: "SummarisedExperiment"
author: "Brandon"
date: "4/26/2021"
output:
  md_document:
    variant: gfm
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

## Creating a SummarisedExperiment Object
1. Creating ColData
```{r cars}
library(SummarizedExperiment)
colData <- data.frame(sample=factor(1:6),
                      condition=factor(c("A","A","B","B","C","C")),
                      treated=factor(rep(0:1,3)))
colData
```

2. Creating RowData or RowRanges
```{r}
library(EnsDb.Hsapiens.v86)
txdb <- EnsDb.Hsapiens.v86
g <- genes(txdb)
g <- keepStandardChromosomes(g, pruning.mode="coarse")
rowRanges <- g[1:10]

```

3. Expression Data/Assay
```{r}
exprs <- matrix(rnorm(6 * 10), ncol=6, nrow=10)

```

4. Combining them into a SE Object
```{r}
se <- SummarizedExperiment(assay=list("exprs"=exprs),
                           colData=colData,
                           rowRanges=rowRanges)
se

```


Inspecting the object
```{r}
assayNames(se)

mcols(se)$score <- rnorm(10)
mcols(se)
```


#Using the ranges of a SummarizedExperiment
    Creating a subset of the se object based on gene location
```{r}
query <- GRanges("1", IRanges(25000,40000))
se.sub <- se[overlapsAny(se, query), ]
se.sub <- se[se %over% query,]
rowRanges(se.sub)
assay(se.sub)
seqinfo(se)
```


# Downloading SummarizedExperiment data from other studies

```{r}
url <- "http://duffel.rail.bio/recount/SRP046226/rse_gene.Rdata"
#download.file(url, "rse_gene.RData")
load("~/GitHub/SummarizedExperiment-Introduction/rse_gene.Rdata")

rse <- rse_gene
```


Inspecting se object
```{r}
colData(rse)[,1:6]

class(rse$characteristics)

rse$characteristics[1:3]

rse$characteristics[[1]]

rse$condition <- sapply(rse$characteristics, `[`, 3)
rse$treatment <- sapply(rse$characteristics, `[`, 4)

rowRanges(rse)

seqinfo(rse)

```


#Visualizing count matrix data in a SummarizedExperiment

creating dds and vsd object

https://www.rdocumentation.org/packages/DESeq2/versions/1.12.3/topics/vst

      This is a wrapper for the varianceStabilizingTransformation (VST) that provides much faster estimation of the dispersion trend used to determine the formula for the VST. 
      The speed-up is accomplished by subsetting to a smaller number of genes in order to estimate this dispersion trend. 
```{r}
library(DESeq2)
library(stringr)
rse@colData@listData$treatment <- str_replace(rse@colData@listData$treatment,":","_")
colData(rse)$condition <- str_replace(colData(rse)$condition,":","_")

dds <- DESeqDataSet(rse, ~ condition + treatment)
vsd <- vst(dds, blind = FALSE)
```

creating count matrix based on conditions
```{r}
library(matrixStats)
rv <- rowVars(assay(vsd))

anno.col <- as.data.frame(colData(vsd)[,c("condition","treatment")])
anno.col
```


plotting heatmap using pheatmap functions
```{r}
library(pheatmap)
pheatmap(assay(vsd)[head(order(rv, decreasing=TRUE),100),],
         annotation_col=anno.col,
         show_rownames=FALSE, show_colnames=FALSE)

sessionInfo()

```

